flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateStreamReader.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult;
+import org.apache.flink.runtime.checkpoint.channel.RefCountingFSDataInputStream.RefCountingFSDataInputStreamFactory;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.BufferBuilder;
+import org.apache.flink.runtime.state.AbstractChannelStateHandle;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import static java.lang.Math.addExact;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateByteBuffer.wrap;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult.HAS_MORE_DATA;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult.NO_MORE_DATA;
+
+/**
+ * Reads the state of a single channel pointed by {@link org.apache.flink.runtime.state.AbstractChannelStateHandle AbstractChannelStateHandle}.
+ * Single-use.
+ * Uses {@link RefCountingFSDataInputStream} internally.
+ */
+@NotThreadSafe
+class ChannelStateStreamReader implements Closeable {
+
+	private final RefCountingFSDataInputStream stream; // todo: buffer?
+	private final Queue<Long> offsets;
+	private final ChannelStateDeserializer serializer;
+	private long pos = -1;
+	private int rem;
+	private boolean closed = false;
+
+	ChannelStateStreamReader(AbstractChannelStateHandle<?> handle, RefCountingFSDataInputStreamFactory streamFactory) {
+		this(streamFactory.forHandle(handle), handle.getOffsets(), streamFactory.getSerializer());
+	}
+
+	private ChannelStateStreamReader(RefCountingFSDataInputStream stream, List<Long> offsets, ChannelStateDeserializer serializer) {
+		this.stream = stream;
+		this.stream.incNumReaders();
+		this.serializer = serializer;
+		this.offsets = new LinkedList<>(offsets);
+	}
+
+	ReadResult readInto(Buffer buffer) throws IOException {
+		return readInto(wrap(buffer));
+	}
+
+	ReadResult readInto(BufferBuilder bufferBuilder) throws IOException {
+		return readInto(wrap(bufferBuilder));
+	}
+
+	private ReadResult readInto(ChannelStateByteBuffer buffer) throws IOException {
+		Preconditions.checkState(!closed, "reader is closed");
+		readWhilePossible(buffer);
+		if (haveMoreData()) {
+			return HAS_MORE_DATA;
+		} else {
+			close();
+			return NO_MORE_DATA;
+		}
+	}
+
+	private void readWhilePossible(ChannelStateByteBuffer buffer) throws IOException {
+		while (haveMoreData() && buffer.isWritable()) {
+			if (pos < 0 || rem <= 0) {
+				advanceOffset();
+			}
+			int bytesRead = serializer.readData(stream, buffer, rem);
+			rem -= bytesRead;
+			pos = addExact(pos, bytesRead);
+		}
+	}
+
+	private boolean haveMoreData() {
+		return rem > 0 || !offsets.isEmpty();
+	}
+
+	private void advanceOffset() throws IOException {
+		//noinspection ConstantConditions
```

Comment:
Not a big fan of leaving inspection settings in code. If everyone does it with different IDEs/settings, it will become quickly a mess.